### ISSUE TITLE
Memory Leak

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#4453]
Migrated from [rt.perl.org#7756](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=7756) (status was 'resolved')

Searchable as RT7756$

**Comment by p5pRT** [comment]
### From pilsl@goldfisch.at



see also discussion at comp\.lang\.perl\.misc with same subject

I experience the bug only at 5\.6\.0 and its vanished at 5\.6\.1\, but another guy says\, he still has the problem at 5\.6\.1 under win\.
Under 5\.6\.0 I have the problem at several linuxes and bsd's\.

Bug&#8203;:
if constructing anonymous structures and inserting anonymous subs via a eval\-statement\, the mem of the whole structure is not released when destroying the structure with undef\.

Workarounds&#8203;: \(that costs time in big structures\)
\- manually undef the nodes holding the anonymous subs before undef the structure
\- manually recursive deletion of the whole structure

severity is high\, cause this is a common contruct in mod\_perl\-environments that makes apache memleaking like a sieve \.\.\.

sample\-script \(just to give an idea\. increase nummber of loops to be sure that is not a free\(\)\-Problem of the running OS\.\) 

\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.
\#\!/usr/bin/perl \-w
use strict;

foreach \(0\.\.0\)   \# extend to increase effect
\{
&nbsp;  allocate\(\);
\}

sub allocate
\{
&nbsp;  print "===============\\n";
&nbsp;  print \`ps \-o vsize\,rss \-p $$ | tail \-1\`;
&nbsp;  my $ptr=\{\};tr\->\{big\}=\[\];
&nbsp;  $\#\{$ptr\->\{big\}\}=5000000;
&nbsp;  def2\($ptr\);
&nbsp;  print &\{$ptr\->\{test\}\->\{func2\}\}\(5\)\,"\\n";
&nbsp;  print \`ps \-o vsize\,rss \-p $$ | tail \-1\`;
&nbsp;  \#  undef $ptr\->\{test\}\->\{func2\};   \# Poss 1 this is enough to kill the whole structure in the next statement
&nbsp;  undef $ptr;                     \# Poss 2 this will not delete the structure without the previous statement
&nbsp;  print \`ps \-o vsize\,rss \-p $$ | tail \-1\`;
\}
sub def2 \{
&nbsp;  my $ptr=shift;
&nbsp;  my $cmd='$ptr\->\{test\}\->\{func2\}=sub\{return $\_\[0\]\+4\};';
&nbsp;  \#  $ptr\->\{test\}\->\{func2\}=sub\{return $\_\[0\]\+4\};        \# Poss A with this statement there is no Problem
&nbsp;  eval $cmd;                                         \# Poss B this statement fools garbage collection
\}
\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.

output&#8203;:  \(shows mainly needed mem by task and the 9 ist just output of the testfunction\)
In the first case the mem is not released \- in the second it is\.
The putput of 'ps' is not a real proove\, but you can increase the loop\.

$\./mem\.pl
 2824 1304
 9
 22360 20904
$vi \# uncomment 19
$\./mem\.pl
===============
 2824 1304
 9
 22360 20904
&nbsp;  2828 1376

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl v5.6.0:

Configured by fg at Wed Apr  4 18:57:12 CEST 2001.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.1-20mdksmp, archname=i386-linux
    uname='linux bi.mandrakesoft.com 2.4.1-20mdksmp #1 smp tue feb 20 19:21:17 cet 2001 i686 unknown '
    config_args='-des -Darchname=i386-linux -Dd_dosuid -Ud_csh -Duseshrplib -Doptimize=-O3 -fomit-frame-pointer -pipe -mcpu=pentiumpro -march=i586 -ffast-math -Dprefix=/usr -Di_ndbm -Di_gdbm -Di_shadow -Di_syslog -Uuselargefiles -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/lib/perl5/man/man3'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=undef 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='cc', optimize='-O3 -fomit-frame-pointer -pipe -mcpu=pentiumpro -march=i586 -ffast-math', gccversion=2.96 20000731 (Linux-Mandrake 8.0 2.96-0.47mdk)
    cppflags='-fno-strict-aliasing'
    ccflags ='-fno-strict-aliasing'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=4
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -ldl -lm -lc -lcrypt
    libc=/lib/libc-2.2.2.so, so=so, useshrplib=true, libperl=libperl.so
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic -Wl,-rpath,/usr/lib/perl5/5.6.0/i386-linux/CORE'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.0:
    /usr/lib/perl5/5.6.0/i386-linux
    /usr/lib/perl5/5.6.0
    /usr/lib/perl5/site_perl/5.6.0/i386-linux
    /usr/lib/perl5/site_perl/5.6.0
    /usr/lib/perl5/site_perl
    .


Environment for perl v5.6.0:
    HOME=/home/pilsl
    LANG=en
    LANGUAGE=en_US:en
    LC_COLLATE=en_US
    LC_CTYPE=en_US
    LC_MESSAGES=en_US
    LC_MONETARY=en_US
    LC_NUMERIC=en_US
    LC_TIME=en_US
    LD_LIBRARY_PATH=/usr/lib
    LOGDIR (unset)
    PATH=/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin:/usr/games:/usr/X11R6/bin:/usr/games
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

This was fixed by commit afa38808e08264de7bcd3b2241ab41424d64d0d4

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'
